common/azureml_appinsights_logger/azureml_appinsights_logger/observability.py

```diff
@@ -0,0 +1,165 @@
+        instance
+        """
+        run = Run.get_context()
+        e = Env()
+        if not run.id.startswith(self.OFFLINE_RUN):
+            print("Registered AML Logger")
+            self.loggers.append(AzureMlLogger(run))
+        if e.app_insights_connection_string:
+            if "InstrumentationKey" in e.app_insights_connection_string:
+                print("Registered AppInsights Logger")
+                self.loggers.append(AppInsightsLogger(run))
+        if e.log_to_console:
+            print("Registered Console Logger")
+            self.loggers.append(ConsoleLogger(run))
+
+
+class Observability(LoggerInterface):
+    _instance = None
+
+    # Straightforward Singleton Pattern from
+    # https://python-patterns.guide/gang-of-four/singleton/
+    def __new__(cls):
+        if cls._instance is None:
+            print('Creating the Observability Singleton')
+            cls._instance = super(Observability, cls).__new__(cls)
+            cls._instance.__initialized = False
+        return cls._instance
+
+    def __init__(self) -> None:
+        if(not self.__initialized):
+            print('Initializing the Observability Singleton')
+            self.__initialized = True
+            self._loggers = Loggers()
+
+    def log_metric(
+            self, name="", value="", description="", log_parent=False,
+    ):
+        """
+        this method sends the metrics to all registered loggers
+        :param name: metric name
+        :param value: metric value
+        :param description: description of the metric
+        :param log_parent: (only for AML), send the metric to the run.parent
+        :return:
+        """
+        for logger in self._loggers.loggers:
+            logger.log_metric(name, value, description, log_parent)
+
+    def log(self, description="", severity=Severity.INFO):
+        """
+        this method sends the logs to all registered loggers
+        :param description: Actual log description to be sent
+        :param severity: log Severity
+        :return:
+        """
+        for logger in self._loggers.loggers:
+            logger.log(description, severity)
+
+    def exception(self, exception: Exception):
+        """
+        this method sends the exception to all registered loggers
+        :param exception: Actual exception to be sent
+        :return:
+        """
+        for logger in self._loggers.loggers:
+            logger.exception(exception)
+
+    def get_logger(self, logger_class):
+        """
+        This method iterate over the loggers and it
+        returns the logger with the same type as the provided one.
+        this is a reference that can be used in case
+        any of the built in functions of the loggers is required
+        :param logger_class:
+        :return: a logger class
+        """
+        for logger in self._loggers.loggers:
+            if type(logger) is type(logger_class):
+                return logger
+
+    def span(self, name='span'):
```

Comment:
As we discussed, the following methods should not be here. 
These should only be the concern of **AppInsightsLogger**.
- span
- start_span
- end_span
- current_span
- add_attribute_to_current_span
- list_collected_spans

Since only AppInsightsLogger support these methods, it's misleading and against our abstraction to be in Observability Class. 

One way to cal this method with Observability is  :
```python
observability.get_logger(AppInsightsLogger("")).start_span(..)
...
observability.get_logger(AppInsightsLogger("")).span(..)
```

Reply:
Thought about this one. I'd prefer to keep the current way because:

- Other loggers could implement span in the future. 
- If multiple loggers implement span, and the caller wants to log to multiple destinations, then they'd need to get_logger and call the logging methods on each one of them, defying the purpose of having a unified logger.

Reply:
While I understand @mokarian concern from a coding perspective, I intentionally put the methods in the Interface that it becomes convenient to use the library as an AML developer. The "heavy load" should be abstracted, even if that results in not getting recorded the expected data.

Also the observability lib/package will automatically activate AppInsights and other loggers based on ENV values and dispatch/chain the calls to the different types of loggers. So on the AML and ML script side you don't need to know about what loggers are or will be active.